server/src/main/java/com/linecorp/centraldogma/server/internal/storage/ScheduledPurgingService.java

```diff
@@ -0,0 +1,252 @@
+import com.linecorp.centraldogma.server.metadata.UserAndTimestamp;
+import com.linecorp.centraldogma.server.storage.project.ProjectManager;
+
+/**
+ * A service class for pursing projects and repositories that were removed before.
+ */
+public class ScheduledPurgingService {
+
+    private static final Logger logger = LoggerFactory.getLogger(ScheduledPurgingService.class);
+
+    /**
+     * How often to run the storage purge schedules. i.e. every minute.
+     */
+    private static final Duration TICK = Duration.ofMinutes(1);
+
+    private final File rootDir;
+    private final ProjectManager projectManager;
+    private final ScheduledExecutorService purgeWorker;
+    private final long maxRemovedRepositoryAgeMillis;
+    private final StoragePurgingScheduler storagePurgingScheduler = new StoragePurgingScheduler();
+
+    public ScheduledPurgingService(
+            File rootDir, ProjectManager projectManager,
+            ScheduledExecutorService purgeWorker, long maxRemovedRepositoryAgeMillis) {
+        this.rootDir = requireNonNull(rootDir, "rootDir");
+        this.projectManager = requireNonNull(projectManager, "projectManager");
+        this.purgeWorker = requireNonNull(purgeWorker, "purgeWorker");
+        checkArgument(maxRemovedRepositoryAgeMillis >= 0,
+                      "maxRemovedRepositoryAgeMillis: %s (expected: >= 0)", maxRemovedRepositoryAgeMillis);
+        this.maxRemovedRepositoryAgeMillis = maxRemovedRepositoryAgeMillis;
+    }
+
+    public void start(CommandExecutor commandExecutor,
+                      MetadataService metadataService) {
+        if (!isDisabled()) {
+            requireNonNull(commandExecutor, "commandExecutor");
+            requireNonNull(metadataService, "metadataService");
+            cleanPurgedFiles();
+            storagePurgingScheduler.start(() -> {
+                try {
+                    schedulePurgeData(commandExecutor, metadataService);
+                } catch (Exception e) {
+                    logger.warn("Unexpected pursing service failure", e);
+                }
+            });
+        }
+    }
+
+    public boolean isStarted() {
+        return storagePurgingScheduler.isStarted();
+    }
+
+    public void stop() {
+        if (!isDisabled()) {
+            storagePurgingScheduler.stop();
+        }
+    }
+
+    private void cleanPurgedFiles() {
+        projectManager.listPurged()
+                      .forEach(projectName -> cleanPurgedFile(projectName, "project"));
+        projectManager.list().forEach((unused, project) -> {
+            project.repos()
+                   .listPurged()
+                   .forEach(repoName -> cleanPurgedFile(repoName, "repository"));
+        });
+    }
+
+    @VisibleForTesting
+    void cleanPurgedFile(String name, String typeName) {
+        final File purged = new File(rootDir, name + SUFFIX_PURGED);
+        CompletableFuture.supplyAsync(() -> {
+            try {
+                Util.deleteFileTree(purged);
```

Comment:
How about:

- Adding some log message around `deleteFileTree()` at INFO just in case it takes long time?
- de-duplicating the same logic in `DirectoryBasedStorageManager`?

Reply:
I also have concerned about duplication logic. 
I made a new interface `void clear(String name)` to permanently delete a specified file.